src/main/java/spoon/support/adaption/DeclarationNode.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * SPDX-License-Identifier: (MIT OR CECILL-C)
+ *
+ * Copyright (C) 2006-2019 INRIA and contributors
+ *
+ * Spoon is available either under the terms of the MIT License (see LICENSE-MIT.txt) of the Cecill-C License (see LICENSE-CECILL-C.txt). You as the user are entitled to choose the terms under which to adopt Spoon.
+ */
+package spoon.support.adaption;
+
+import spoon.SpoonException;
+import spoon.reflect.declaration.CtType;
+import spoon.reflect.declaration.CtTypeParameter;
+import spoon.reflect.reference.CtTypeParameterReference;
+import spoon.reflect.reference.CtTypeReference;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+
+import static spoon.support.adaption.NodePrintHelper.quote;
+import static spoon.support.adaption.NodePrintHelper.toJsonLikeArray;
+
+class DeclarationNode {
+
+	private final List<CtTypeParameter> formalArguments;
+	private final CtTypeReference<?> inducedBy;
+	private final Set<GlueNode> lowerNodes;
```

Comment:
> but as far as I can tell, the same (identity-wise and content-wise) node cannot be added twice anyway

That is precisely the reason why I made it a set. I don't need any order, I don't care about index based access and I do not intend do ever add duplicates. It just sounds like a good usecase for a set to me.

In its current form it does not guard against creating multiple nodes for the same reference as you noticed, so this is mostly of symbolic value. I am not sure I agree with it, but I will change it to a list.

Reply:
> That is precisely the reason why I made it a set. I don't need any order, I don't care about index based access and I do not intend do ever add duplicates. It just sounds like a good usecase for a set to me.

I would have to disagree on that. Using a set indicates that uniqueness (and consequently deduplication of "careless" insertions) is semantically important.

Using an explicitly identity-based set here would be mildly misleading as it hints that there is some form of importance to the object identities (which there is not), but it would at least be clear that the set operates on identity.

Using a `HashSet` might be significantly misleading as someone coming in later to edit the code might assume that the objects contained in the `HashSet` have some concept of equality beyond identity. I would.

There's also the fact that the primary use case for the child list is to iterate over it. Operations on a `HashSet` (most prominently inserting and iterating) are several times slower than on an `ArrayList`. Compounding all of that, I think a list is a much better choice.

> In its current form it does not guard against creating multiple nodes for the same reference as you noticed, so this is mostly of symbolic value.

This is precisely what is misleading to me.

Reply:
> Using a set indicates that uniqueness (and consequently deduplication of "careless" insertions) is semantically important

I agree with the first part and that is the case here, bit I never considered the part in parenthesis to be of relevance when picking a collection. Interesting, I will keep that in mind when choosing something next time.

A general `Collection` randomly instantiated to an `ArrayList` (as that is the "default" collection in most cases) might be a better solution then, but it also isn't really important. It is declared as a `List` now, that should be good enough. Most people do not necessarily think "is in a certain order" when they see `List` as much as they think "must be unique" when seeing a `Set`.

Reply:
> A general Collection randomly instantiated to an ArrayList

Yes, I agree that would be even better in flagging the the point of "this is just a collection with no particular characteristics to rely on". If I had thought of that I would have made that my suggestion :)